PS-9215 Update telemetry information (#319)
percona/psmysql-docs: docs/telemetry.md

* The Telemetry Agent is an independent process running on your database host's operating system and carries out the following tasks:

    * Collects OS-level metrics

    * Reads the Metrics file, adds the OS-level metrics

    * Sends the full set of metrics to the Percona Platform

The telemetry also uses the Percona Platform with the following components:

* Telemetry Service - offers an API endpoint for sending telemetry. The service handles incoming requests. This service saves the data into Telemetry Storage.

* Telemetry Storage - stores all telemetry data for the long term.

## DB component

The DB component extends the functionality of the database. This functionality can be implemented in the following ways, depending on the database:

* Modifying the source code

* Add modular components

* Add self-contained extensions

In Percona Server for MySQL, we have a DB component. The DB component is enabled by default during the initial database deployment. 

This component does not collect the following information:

* database names

* user names or credentials

* any user-entered value

> The component sends metrics gathered from the database instance to the Metrics file. This information is collected, by default, every 24 hours. Each connection creates a new Metrics file. When the file date exceeds seven days, the component removes the outdated files before creating the new file.

Reviewer: remove "by default" maybe. User has no (official) way to change it.

> The component creates a file in the local file system using a timestamp as the name with a `.tmp` extension, transfers the metrics to this file, and then renames the extension to .json.

Reviewer: No need to mention about tmp files. It is implementation detail. Mention just about json files.

## Metrics file

Percona stores the Metrics file in one of the following directories on the local file system. The location depends on the product.

* Telemetry root path - `/usr/local/percona/telemetry`

* PSMDB root path -   `${telemetry root path}/psmdb/`

* PS root path -   `${telemetry root path}/ps/`

* PXC root path - `${telemetry root path}/pxc/`

* PG root path - `${telemetry root path}/pg/`

* Percona ToolKit - `${telemetry root path}/toolkit/`

Percona archives the telemetry history in `${telemetry root path}/history/`.

### Metrics file format

> The Metrics file uses the Javascript Object Notation (JSON) format, structured in key-value pairs with a colon separating them (:).

Reviewer: Remove everything after coma ("structured in ... ")

> * Keys are always strings enclosed in double quotes ("db_instance_id").

Reviewer: I would remove this and following details up to "Percona reserves...."

* Values can be in the following data types:

    * String (enclosed in double quotes)

    * Number (integer or floating point)

    * Booleans (true or false)

    * Null (the absence of a value)

    * Arrays (ordered collection of values)

    * Objects (unordered collections of key-value pairs)

Objects are enclosed in curly braces ({}), and arrays are enclosed in square brackets ([]). Whitespace is ignored, except when part of a string value.

Percona reserves the right to extend the current set of JSON structure attributes in the future.  

> The top-level keys are the following:

Reviewer: Do we need to describe these details?

|--------------------|-------------|---------------------|----------------------------------------------------------------------------------------------------------|
| `db_instance_id`   | Mandatory   | DB plugin only     | This key is the ID of the componenticular DB instance. It is used for all telemetry results from the same database instance and is saved between instance restarts. |
| `db_cluster_id`    | Optional    | DB plugin only     | This key is the ID of componenticular DB instances grouped as DB cluster. It can be taken from the database instance or generated and is used for all telemetry results within the same cluster. It is saved between database instance restarts. |
| `db_replication_id`| Optional    | DB plugin only     | This key is the ID of a specific DB replication setup (Main→secondary). It can be taken from the database instance or generated and is used for all telemetry results within the same replication setup. It is saved between database instance restarts. |
| `pillar_version`   | Mandatory   | DB plugin only     | This key identifies the exact version of Percona product. It is used for backward compatibility with the earlier Telemetry project. |

> An example of the data collected is the following:

Reviewer: I would include only example, without above details about what is mandatory, what is not

| Identify Issues Early       | Telemetry can catch bugs or performance problems before they become widespread. |

Benefits for Users in the Long Run:

| Advantages                | Description                                                                                                         |
|---------------------------|---------------------------------------------------------------------------------------------------------------------|
| Faster Bug Fixes          | With telemetry data, developers can pinpoint issues affecting specific users and prioritize fixing them quickly.  |
| Better Features           | Telemetry helps developers understand user needs and preferences. This allows them to focus on features that will be genuinely useful and improve your overall experience. |
| A More Stable Experience  | By identifying and resolving issues early, telemetry helps create a more stable and reliable software experience for everyone. |

## Installation-time telemetry

> This telemetry is executed only once during software installation. It collects information at the moment of installation and does not run afterward.

Reviewer 2: ```suggestion
This telemetry is executed only once during software installation or Docker container startup. It collects information at the moment of installation and does not run afterward.
```

# Telemetry on Percona Server for MySQL

Percona wants to learn more about how people use their software. This telemetry feature helps them understand what features are used the most, if there are any problems, and how they can improve their software in the future.

You can choose not to share this information. Participation in this program is completely voluntary. If you're uncomfortable sharing this anonymous data, you can [disable telemetry].

> Your privacy is protected. Percona doesn't collect any personal information about you. All the data they collect is anonymous, meaning it can't be traced back to any individual user.

Reviewer 2: ```suggestion
Your privacy is protected. Percona doesn't collect any personal information about you. All the data they collect is anonymous, meaning it can't be traced back to any individual user or host/server.
```

We have a Percona Privacy statement explaining how we handle your data in more detail.

Right now, this telemetry feature is only included in software packages and Docker images.

## Telemetry components

Percona collects information using the following components:

* The Percona Telemetry DB part collects the necessary metrics directly from the database and stores them in a Metrics file.

* The Metrics file stores the metrics and is a standalone file located on the database host's file system.

* The Telemetry Agent is an independent process running on your database host's operating system and carries out the following tasks:

    * Collects OS-level metrics

>     * Reads the Metrics file, adds the OS-level metrics

Reviewer 2: And it collects the list of installed Percona's packages via local package manager

    * Sends the full set of metrics to the Percona Platform

The telemetry also uses the Percona Platform with the following components:

* Telemetry Service - offers an API endpoint for sending telemetry. The service handles incoming requests. This service saves the data into Telemetry Storage.

* Telemetry Storage - stores all telemetry data for the long term.

## DB component

The DB component extends the functionality of the database. This functionality can be implemented in the following ways, depending on the database:

* Modifying the source code

* Add modular components

* Add self-contained extensions

In Percona Server for MySQL, we have a DB component. The DB component is enabled by default during the initial database deployment. 

This component does not collect the following information:

* database names

* user names or credentials

* any user-entered value

> The component sends metrics gathered from the database instance to the Metrics file. This information is collected every 24 hours. Each connection creates a new Metrics file. When the file date exceeds seven days, the component removes the outdated files before creating the new file.

Reviewer 3: > Each connection creates a new Metrics file.

Maybe 'Each collection'?

The component creates a file in the local file system using a timestamp as the name with a `.json` extension.

## Metrics file

Percona stores the Metrics file in one of the following directories on the local file system. The location depends on the product.

* Telemetry root path - `/usr/local/percona/telemetry`

> * PSMDB root path -   `${telemetry root path}/psmdb/`

Reviewer 3: PSMDB will have 2 paths:
`- PSMDB (mongod) root path -   `${telemetry root path}/psmdb/`
- PSMDB (mongos) root path -   `${telemetry root path}/psmdbs/``

* PS root path -   `${telemetry root path}/ps/`

* PXC root path - `${telemetry root path}/pxc/`

* PG root path - `${telemetry root path}/pg/`

> * Percona ToolKit - `${telemetry root path}/toolkit/`

Reviewer 3: toolkit was not implemented.

# Telemetry on Percona Server for MySQL

Percona wants to learn more about how people use their software. This telemetry feature helps them understand what features are used the most, if there are any problems, and how they can improve their software in the future.

You can choose not to share this information. Participation in this program is completely voluntary. If you're uncomfortable sharing this anonymous data, you can [disable telemetry].

Your privacy is protected. Percona doesn't collect any personal information about you. All the data they collect is anonymous, meaning it can't be traced back to any individual user or host/server.

We have a Percona Privacy statement explaining how we handle your data in more detail.

> Right now, this telemetry feature is only included in software packages and Docker images.

Reviewer 3: tarballs also have it. Please add

The component sends metrics gathered from the database instance to the Metrics file. This information is collected every 24 hours. Each collection creates a new Metrics file. When the file date exceeds seven days, the component removes the outdated files before creating the new file.

The component creates a file in the local file system using a timestamp as the name with a `.json` extension.

## Metrics file

Percona stores the Metrics file in one of the following directories on the local file system. The location depends on the product.

* Telemetry root path - `/usr/local/percona/telemetry`

* PSMDB (mongod) root path -   `${telemetry root path}/psmdb/`

* PSMDB (mongos) root path - `${telemetry root path}/psmdbs/'

* PS root path -   `${telemetry root path}/ps/`

* PXC root path - `${telemetry root path}/pxc/`

* PG root path - `${telemetry root path}/pg/`

Percona archives the telemetry history in `${telemetry root path}/history/`.

### Metrics file format

The Metrics file uses the Javascript Object Notation (JSON) format. Percona reserves the right to extend the current set of JSON structure attributes in the future.  

An example of the data collected is the following:

> 

Reviewer 3: This is invalid example of metrics file. Here is a correct one (exectly metrics, not history):
`{
    "db_instance_id": "e83c568c-e140-11ee-8320-7e207666b18a",
    "pillar_version": "8.0.35-27",
    "active_plugins": [
        "binlog",
        "mysql_native_password",
        "sha256_password",
        "caching_sha2_password",
        "sha2_cache_cleaner",
        "daemon_keyring_proxy_plugin",
        "PERFORMANCE_SCHEMA",
        "CSV",
        "MEMORY",
        "InnoDB",
        "INNODB_TRX",
        "INNODB_CMP",
        "INNODB_CMP_RESET",
        "INNODB_CMPMEM",
        "INNODB_CMPMEM_RESET",
        "INNODB_CMP_PER_INDEX",
        "INNODB_CMP_PER_INDEX_RESET",
        "INNODB_BUFFER_PAGE",
        "INNODB_BUFFER_PAGE_LRU",
        "INNODB_BUFFER_POOL_STATS",
        "INNODB_TEMP_TABLE_INFO",
        "INNODB_METRICS",
        "INNODB_FT_DEFAULT_STOPWORD",
        "INNODB_FT_DELETED",
        "INNODB_FT_BEING_DELETED",
        "INNODB_FT_CONFIG",
        "INNODB_FT_INDEX_CACHE",
        "INNODB_FT_INDEX_TABLE",
        "INNODB_TABLES",
        "INNODB_TABLESTATS",
        "INNODB_INDEXES",
        "INNODB_TABLESPACES",
        "INNODB_COLUMNS",
        "INNODB_VIRTUAL",
        "INNODB_CACHED_INDEXES",
        "INNODB_SESSION_TEMP_TABLESPACES",
        "MyISAM",
        "MRG_MYISAM",
        "TempTable",
        "ARCHIVE",
        "BLACKHOLE",
        "ngram",
        "mysqlx_cache_cleaner",
        "mysqlx",
        "ROCKSDB",
        "rpl_semi_sync_source",
        "ROCKSDB_CFSTATS",
        "ROCKSDB_DBSTATS",
        "ROCKSDB_PERF_CONTEXT",
        "ROCKSDB_PERF_CONTEXT_GLOBAL",
        "ROCKSDB_CF_OPTIONS",
        "ROCKSDB_GLOBAL_INFO",
        "ROCKSDB_COMPACTION_HISTORY",
        "ROCKSDB_COMPACTION_STATS",
        "ROCKSDB_ACTIVE_COMPACTION_STATS",
        "ROCKSDB_DDL",
        "ROCKSDB_INDEX_FILE_MAP",
        "ROCKSDB_LOCKS",
        "ROCKSDB_TRX",
        "ROCKSDB_DEADLOCK"
    ],
    "active_components": [
        "file://component_percona_telemetry"
    ],
    "uptime": "6185",
    "databases_count": "7",
    "databases_size": "33149",
    "se_engines_in_use": [
        "InnoDB",
        "ROCKSDB"
    ],
    "replication_info": {
        "is_semisync_source": "1",
        "is_replica": "1"
    }
}`

{"key" : "hardware_arch","value" : "x86_64 x86_64"}]}]
```

### Disable installation telemetry

This telemetry feature is enabled by default.

You can disable this telemetry if you decide not to send installation data to Percona. Set the `PERCONA_TELEMETRY_DISABLE=1` environment variable for either the root user or the operating system before installing.

These actions do not affect the continuous telemetry system.

=== "Debian-derived distribution"

    Add the environment variable before the installation process.

    ```{.bash data-prompt="$"}
    $ sudo PERCONA_TELEMETRY_DISABLE=1 apt install percona-server-server
    ```

=== "Red Hat-derived distribution"

    Add the environment variable before the installation process.
    
    ```{.bash data-prompt="$"}
    $ sudo PERCONA_TELEMETRY_DISABLE=1 yum install percona-server-server
    ```

=== "DOCKER"

    Add the environment variable when running a command in a new container.
    
    ```{.bash data-prompt="$"}

>     $ docker run -d -e MYSQL_ROOT_PASSWORD=test1234# -e PERCONA_TELEMETRY_DISABLE=1 -e --name=percona-server percona/percona-server:8.1

Reviewer 2: ```suggestion
    $ docker run -d -e MYSQL_ROOT_PASSWORD=test1234# -e PERCONA_TELEMETRY_DISABLE=1 --name=percona-server percona/percona-server:8.1
```

    ```

## Continuous telemetry

This telemetry system involves setting up a telemetry agent and a database (DB) component. It continuously collects and sends information daily.
 

>  The telemetry agent runs continuously or at scheduled daily intervals to collect data. The agent gathers data (for example, usage statistics) and sends this information to the Percona platform.

Reviewer 2: ```suggestion
 The telemetry agent runs at scheduled daily intervals to collect data. The agent gathers data (for example, usage statistics) and sends this information to the Percona platform.
```

Percona includes the telemetry systems only in software packages, compressed archives (tarballs), and Docker images. The continuous telemetry system requires a telemetry agent and a specific folder to write data when installed from compressed archives (tarballs). If these elements are available, telemetry collects and sends the data.

## Elements of the continuous telemetry system

Percona collects information using these elements:

| Function                          | Description |
|-------------------------------|-------------|

> | [Percona Telemetry DB component](#percona-telemetry-agent) | This component collects metrics directly from the database and stores them in a metrics file. |

Reviewer 2: ```suggestion
| [Percona Telemetry DB component](#overview-of-the-db-component) | This component collects metrics directly from the database and stores them in a metrics file. |
```

| [Metrics File](#locations-of-metrics-files-and-telemetry-history) | This standalone file on the database host's file system stores the collected metrics. |

> | [Telemetry Agent](#how-the-percona-telemetry-agent-operates) | This independent process runs on your database host's operating system and performs the following tasks:<br> - Collects OS-level metrics.<br>- Reads the metrics file and adds the OS-level metrics.<br>- Sends the complete set of metrics to the Percona Platform.<br> - Collects the list of installed Percona packages using the local package manager. |

Reviewer 2: ```suggestion
| [Telemetry Agent](#percona-telemetry-agent) | This independent process runs on your database host's operating system and performs the following tasks:<br> - Collects OS-level metrics.<br>- Reads the metrics file and adds the OS-level metrics.<br>- Sends the complete set of metrics to the Percona Platform.<br> - Collects the list of installed Percona packages using the local package manager. |
```

| Telemetry Storage | This component stores all telemetry data for the long term.                                  |

## Overview of the DB component

The Percona Server for MySQL includes a DB component by default during installation. This component extends the database's functionality in one of these ways, depending on the specific database:

* Modifying the source code directly

* Adding modular components

* Adding self-contained extensions

The DB component has the following qualities:

* Collects metrics from the database daily

* Writes these metrics to a new JSON file on the local file system, named with a timestamp and the .json extension

* Stores only the most recent week's data by deleting older Metrics files before creating a new one.

The DB component does NOT collect the following:

* Database names

* User names or credentials

* Data entered by users

## Locations of metrics files and telemetry history

Percona stores the Metrics file in one of the following directories on the local file system. The location depends on the product.

* Telemetry root path - `/usr/local/percona/telemetry`

* PSMDB (mongod) root path -   `${telemetry root path}/psmdb/`

> * PSMDB (mongos) root path - `${telemetry root path}/psmdbs/'

Reviewer 2: ```suggestion
* PSMDB (mongos) root path - `${telemetry root path}/psmdbs/`
```

* PS root path -   `${telemetry root path}/ps/`

* PXC root path - `${telemetry root path}/pxc/`

* PG root path - `${telemetry root path}/pg/`

Percona archives the telemetry history in `${telemetry root path}/history/`.

### Metrics file format

The Metrics file uses the Javascript Object Notation (JSON) format. Percona reserves the right to extend the current set of JSON structure attributes in the future.  

> An example of the data collected is the following:

Reviewer 2: ```suggestion
An example of the Metrics file content is the following:
```

In the first 24 hours, no information is collected or sent. After that period, the agent tries to send the collected information to Percona's servers (Percona Platform) daily. If this operation fails, the agent retries up to five times. After the data is successfully sent, the agent saves a copy of the sent data in a separate "history" folder, and then, deletes the original file created by the database.

The agent won't send any data if the target directory doesn't contain specific files related to Percona software.

### Telemetry agent payload example

The following is an example of a telemetry agent payload:

```json
{
  "reports": [
    {
      "id": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BB",
      "createTime": "2023-09-01T10:56:49Z",
      "instanceId": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BA",
      "productFamily": "PRODUCT_FAMILY_PS",
      "metrics": [
        {
          "key": "OS",
          "value": "Ubuntu"
        },
        {
          "key": "pillar_version",
          "value": "8.0.33-25"
        }
      ]
    }
  ]
}
```

The agent sends information about the database and metrics.

| Key | Description |
|---|---|

> | "id" | A generated Universally Unique Identifier (UUID) version 4 |

Reviewer 2: ```suggestion
| "id" | A randomly generated Universally Unique Identifier (UUID) version 4 of request |
```

| "createTime" | UNIX timestamp |
| "instanceId" | The DB Host ID. The value can be taken from the `instanceId`, the `/usr/local/percona/telemetry_uuid` or generated as a UUID version 4 if the file is absent. |
| "productFamily" | The value from the file path |
| "metrics" | An array of key:value pairs collected from the Metrics file.

The following operating system-level metrics are sent with each check:

| Key | Description |
|---|---|
| "OS" | The name of the operating system |

> | "hardware_arch" | The type of processed used in the environment |

Reviewer 2: ```suggestion
| "hardware_arch" | CPU architecture used on DB host |
```

{"key" : "hardware_arch","value" : "x86_64 x86_64"}]}]
```

> ## Telemetry Agent

Reviewer 3: I am not sure, but maybe we should add that Telemetry Agent is delivered as a mandatory dependency to PS packages? And if smb tries to delete TA package - the DB will also be deleted?

+ The telemetry will also be in tarballs. If there is no Telemetry Agent and dedicated folder, it will simply fail to be written, if there are - they will be written and sent.

Reviewer: I don't think it will be so easy to uninstall TA package. Package manager will complain that it is a dependency of DB and will not take action.

Reviewer 3: But smb can try autoremove on deb, it automatically removes dependencies.

Reviewer: TA is the dependency of DB. AFAIK if you remove DB with autoremove, it will remove TA, not the other way. But someone smarter than me in this matter should give the opinion here :). @EvgeniyPatlan

Reviewer 3: Kamil,

As TA is hard/mandatory dependency you can not delete it without deleting dependent package. I was even a bit mistaken, you do not need autoremove, simple remove does the job:

`vagrant@d-buster-64-install:~$ sudo apt list --installed|grep percona

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

percona-release/stable,now 1.0-29.generic all [installed]
percona-server-client/unknown,now 8.0.36-28-1.buster amd64 [installed,automatic]
percona-server-common/unknown,now 8.0.36-28-1.buster amd64 [installed,automatic]
percona-server-server/unknown,now 8.0.36-28-1.buster amd64 [installed]
percona-telemetry-agent/stable,now 1.0.0-1.buster amd64 [installed,automatic]

vagrant@d-buster-64-install:~$ sudo apt remove percona-telemetry-agent -y
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  libmecab2 libnuma1 percona-server-client percona-server-common
Use 'sudo apt autoremove' to remove them.
!!!!!!!! The following packages will be REMOVED:
!!!!!!!!  percona-server-server percona-telemetry-agent
0 upgraded, 0 newly installed, 2 to remove and 80 not upgraded.
After this operation, 614 MB disk space will be freed.
(Reading database ... 30182 files and directories currently installed.)
Removing percona-server-server (8.0.36-28-1.buster) ...
update-alternatives: removing manually selected alternative - switching my.cnf to auto mode
update-alternatives: using /etc/mysql/my.cnf.fallback to provide /etc/mysql/my.cnf (my.cnf) in auto mode
Removing percona-telemetry-agent (1.0.0-1.buster) ...
Stopping Percona Telemetry Agent service...
Processing triggers for man-db (2.8.5-2) ...
Processing triggers for libc-bin (2.28-10+deb10u2) ...

vagrant@d-buster-64-install:~$ sudo apt list --installed|grep percona

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

percona-release/stable,now 1.0-29.generic all [installed]
percona-server-client/unknown,now 8.0.36-28-1.buster amd64 [installed,auto-removable]
percona-server-common/unknown,now 8.0.36-28-1.buster amd64 [installed,auto-removable]

> The Percona Telemetry agent functions as a dedicated OS daemon process. The service name on the operating system is `percona-telemetry-agent`. The agent can create, read, write, and delete files in the `${telemetry root path}` and only interacts with JSON files.

Reviewer 3: Please add that agent log file can be found at `/var/log/percona/telemetry-agent.log` (this is for both rpm and deb)

> The agent sends the collected information as a file to the Percona Platform every 24 hours. If the attempt fails, the agent tries again after a defined period with a limit of five attempts. After a successful transmission, the agent deletes the file.

Reviewer 3: After a successful transmission TA saves file with all the sent metrics to history folder. Pillar metrics file is deleted.

      "createTime": "2023-09-01T10:56:49Z",
      "instanceId": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BA",
      "productFamily": "PRODUCT_FAMILY_PS",
      "metrics": [
        {
          "key": "OS",
          "value": "Ubuntu"
        },
        {
          "key": "pillar_version",
          "value": "8.0.33-25-1"
        }
      ]
    }
  ]
}
```

The agent sends information about the database and metrics.

| Key | Description |
|---|---|
| "id" | A generated Universally Unique Identifier (UUID) version 4 |
| "createTime" | UNIX timestamp |
| "instanceId" | The DB Host ID. The value can be taken from the `instanceId`, the `/usr/local/percona/telemetry_uuid` or generated as a UUID version 4 if the file is absent. |
| "productFamily" | The value from the file path |
| "metrics" | An array of key:value pairs collected from the Metrics file.

The following operating system-level metrics are sent with each check:

| Key | Description |
|---|---|
| "OS" | The name of the operating system |
| "hardware_arch" | The type of processed used in the environment |
| "deployment" | How the application was deployed. <br> The possible values could be "PACKAGE" or "DOCKER". |

> | "installed_packages" | A list of the installed Percona packages.|

Reviewer 3: For PG we send not only Percona packages.

The information includes the following:

* Package name

* Package version - the same format as Red Hat Enterprise Linux or Debian

* Package repository - if possible

The package names must fit the following pattern:

* `percona-*`

* `Percona-*`

* `proxysql*`

* `pmm`

* `etcd*`

* `haproxy`

* `patroni`

* `pg*`

* `postgis`

* `wal2json`

## Disable telemetry

Percona software you just installed collects information about how you use it (telemetry) by default. Here's how to turn it off:

### Within the First 24 Hours

This is the easiest option. You have 24 hours to disable telemetry before any information is collected or sent.

> Use the following command to turn it off:

Reviewer 3: After restart telemetry-agent will be enabled again. To permanently disable it it should also be disabled:
`systemctl disable percona-telemetry-agent`

```{.bash data-prompt=$}
$ systemctl stop percona-telemetry-agent
```

Even after stopping the telemetry agent service, a different part of the software (DB component) continues to create a file related to telemetry every day and saves that file for seven days.

### To stop file creation

You'll need to uninstall this component using the following command (without restarting the database server):

```{.bash data-prompt="$"}
$ UNINSTALL COMPONENT "file://component_percona_telemetry";
```

If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, it will be enabled again.

### To permanently disable telemetry

You must edit the my.cnf configuration file and add the following line:

```{text}
percona_telemetry_disable=1
```

This line tells the software to keep telemetry off.

### Disable in Docker

> If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, then telemetry is enabled again.

Reviewer 3: Should this be in Docker part?

      "createTime": "2023-09-01T10:56:49Z",
      "instanceId": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BA",
      "productFamily": "PRODUCT_FAMILY_PS",
      "metrics": [
        {
          "key": "OS",
          "value": "Ubuntu"
        },
        {
          "key": "pillar_version",
          "value": "8.0.33-25"
        }
      ]
    }
  ]
}
```

The agent sends information about the database and metrics.

| Key | Description |
|---|---|
| "id" | A generated Universally Unique Identifier (UUID) version 4 |
| "createTime" | UNIX timestamp |
| "instanceId" | The DB Host ID. The value can be taken from the `instanceId`, the `/usr/local/percona/telemetry_uuid` or generated as a UUID version 4 if the file is absent. |
| "productFamily" | The value from the file path |
| "metrics" | An array of key:value pairs collected from the Metrics file.

The following operating system-level metrics are sent with each check:

| Key | Description |
|---|---|
| "OS" | The name of the operating system |
| "hardware_arch" | The type of processed used in the environment |
| "deployment" | How the application was deployed. <br> The possible values could be "PACKAGE" or "DOCKER". |

> | "installed_packages" | A list of the installed Percona packages. <br> Other information is sent on a PostgreSQL installation. |

Reviewer 2: ```suggestion
| "installed_packages" | A list of the installed Percona packages. |
```

* `postgis`

* `wal2json`

## Disable telemetry

Percona software you just installed collects information about how you use it (telemetry) by default. Here's how to turn it off:

### Within the First 24 Hours

You have 24 hours to disable telemetry before any information is collected or sent.

If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, then telemetry is enabled again.

Use the following command to turn it off until the next restart:

```{.bash data-prompt=$}
$ systemctl stop percona-telemetry-agent
```

Even after stopping the telemetry agent service, a different part of the software (DB component) creates a file related to telemetry every day and saves that file for seven days.

To permanently disable telemetry, use this command:

```{.bash data-prompt=$}
$ systemctl disable percona-telemetry-agent
```

### To stop file creation

You'll need to uninstall this component using the following command (without restarting the database server):

```{.bash data-prompt="$"}

> $ UNINSTALL COMPONENT "file://component_percona_telemetry";

Reviewer 2: I think it is worth to mention that this CMD shall be executed in PS console, because the previous commands were executed in Linux shell

```

If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, it will be enabled again.

### To permanently disable telemetry

You must edit the my.cnf configuration file and add the following line:

```{text}

> percona_telemetry_disable=1

Reviewer 2: Shall it be in some sections in my.cnf ?

{"key" : "hardware_arch","value" : "x86_64 x86_64"}]}]
```

## Telemetry Agent

The Percona Telemetry agent functions as a dedicated OS daemon process. The service name on the operating system is `percona-telemetry-agent`. The agent can create, read, write, and delete files in the `${telemetry root path}` and only interacts with JSON files.

You can find the Telemetry agent log file at `/var/log/percona/telemetry-agent.log`.

> The Telemetry agent is a mandatory dependency to Percona Server for MySQL packages. Deleting the agent also deletes the database. 

Reviewer 3: Please also add re packages removal if TA is installed as a dependency (this is based on packaging system behavior not smth special that we did):
- On RPM systems the TA package should be automatically removed when the last dependency package is removed.
- On deb-based systems (Ubuntu and Debian) the Telemetry Agent will **NOT** be automatically deleted when the DB package is removed. To delete Telemetry Agent together with DB the `--autoremove` should be passed but this automatically removes all the dependencies. ( https://perconadev.atlassian.net/browse/PKG-34 ).

The same applies to Centos 7 (yum autoremove should be used) but I am not sure whether we will release on Centos7 or no.

| "productFamily" | The value from the file path |
| "metrics" | An array of key:value pairs collected from the Metrics file.

The following operating system-level metrics are sent with each check:

| Key | Description |
|---|---|
| "OS" | The name of the operating system |
| "hardware_arch" | The type of processed used in the environment |
| "deployment" | How the application was deployed. <br> The possible values could be "PACKAGE" or "DOCKER". |
| "installed_packages" | A list of the installed Percona packages. |

The information includes the following:

* Package name

* Package version - the same format as Red Hat Enterprise Linux or Debian

* Package repository - if possible

The package names must fit the following pattern:

* `percona-*`

* `Percona-*`

* `proxysql*`

* `pmm`

* `etcd*`

* `haproxy`

* `patroni`

* `pg*`

* `postgis`

* `wal2json`

## Disable telemetry

Percona software you just installed collects information about how you use it (telemetry) by default. Here's how to turn it off:

### Within the First 24 Hours

You have 24 hours to disable telemetry before any information is collected or sent.

> If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, then telemetry is enabled again.

Reviewer 3: This seems to be in the wrong place? It is also below

{"key" : "hardware_arch","value" : "x86_64 x86_64"}]}]
```

## Telemetry Agent

The Percona Telemetry agent functions as a dedicated OS daemon process. The service name on the operating system is `percona-telemetry-agent`. The agent can create, read, write, and delete files in the `${telemetry root path}` and only interacts with JSON files.

The agent sends the collected information as a file to the Percona Platform every 24 hours. If the attempt fails, the agent tries again after a defined period with a limit of five attempts. After a successful transmission, the agent deletes the file.

> The agent does not send anything if no Percona-specific files are in the target directory.

Reviewer: Is that true? No OS-level metrics sent?

Author: That's what is described in the project document.

The following is an example of a telemetry agent payload:

```json
{
  "reports": [
    {
      "id": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BB",
      "createTime": "2023-09-01T10:56:49Z",
      "instanceId": "B5BDC47B-B717-4EF5-AEDF-41A17C9C18BA",
      "productFamily": "PRODUCT_FAMILY_PS",
      "metrics": [
        {
          "key": "OS",
          "value": "Ubuntu"
        },
        {
          "key": "pillar_version",
          "value": "8.0.33-25-1"
        }
      ]
    }
  ]
}
```

The agent sends information about the database and metrics.

| Key | Description |
|---|---|
| "id" | A generated Universally Unique Identifier (UUID) version 4 |
| "createTime" | UNIX timestamp |

> | "instanceId" | The DB Host ID. The value can be taken from the `instanceId`, the `/usr/loca/percona/telemetry_uuid` or generated as a UUID version 4 if the file is absent. |

Reviewer 2: ```suggestion
| "instanceId" | The DB Host ID. The value can be taken from the `instanceId`, the `/usr/local/percona/telemetry_uuid` or generated as a UUID version 4 if the file is absent. |
```

Author: Removed based on Kamil's suggestion.

| "productFamily" | The value from the file path |
| "metrics" | An array of key:value pairs collected from the Metrics file.

The following operating system-level metrics are sent with each check:

> | Key | Description |
> |---|---|
> | "OS" | The name of the operating system |
> | "hardware_arch" | The type of processed used in the environment |
> | "deployment" | How the application was deployed. <br> The possible values could be "PACKAGE" or "DOCKER". |

Reviewer 2: ```suggestion
| Key | Description |
|---|---|
| "OS" | The name of the operating system |
| "hardware_arch" | The type of processed used in the environment |
| "deployment" | How the application was deployed. <br> The possible values could be "PACKAGE" or "DOCKER". |
| "installed_packages" | List of installed Percona's packages. |
```

A list of the installed Percona packages. The information includes the following:

* Package name

* Package version - the same format as Red Hat Enterprise Linux or Debian

* Package repository - if possible

The package names must fit the following pattern:

* `percona-*`

* `Percona-*`

* `proxysql*`

* `pmm`

* `etcd*`

* `haproxy`

* `patroni`

* `pg*`

* `postgis`

* `wal2json`

## Disable telemetry

Percona software you just installed collects information about how you use it (telemetry) by default. Here's how to turn it off:

### Within the First 24 Hours

This is the easiest option. You have 24 hours to disable telemetry before any information is collected or sent.

Use the following command to turn it off:

```{.bash data-prompt=$}
$ systemctl stop percona-telemetry-agent
```

Even after stopping the telemetry agent service, a different part of the software (DB component) continues to create a file related to telemetry every day and saves that file for seven days.

> ### To stop file creation

Reviewer: This instruction is only for PS/PXC

You'll need to uninstall this component using the following command (without restarting the database server):

```{.bash data-prompt="$"}
$ UNINSTALL COMPONENT "file://component_percona_telemetry";
```

If you restart the database server after running the uninstall command, the telemetry component might automatically turn back on. This event happens because the server reloads settings during restart, and if those settings include instructions to use telemetry, it will be enabled again.

### To permanently disable telemetry

You must edit the my.cnf configuration file and add the following line:

```{text}

> percona_telemetry_disable-1

Reviewer: =1
